lib/utils/filter_samples.py

```diff
@@ -0,0 +1,117 @@
+            RETURN {{ [ meta.ok ]: {{ [ meta.k ]: meta.v }} }}
+        )
+        return {{"id": node.id, "version": node.ver, "meta": APPLY("MERGE_RECURSIVE", meta_objs_values)}}
+)
+for node in node_metas
+    FILTER """
+
+class SampleFilterer():
+    '''
+    '''
+    def __init__(cls, ctx, re_api_url, sample_service):
+        cls.re_api_url = re_api_url
+        cls.sample_service = sample_service
+        cls.re_admin_token = ctx.get('token')  # needs to be a env/config variable
+
+    def filter_samples(self, params):
+        samples, filter_conditions = parse_input(params)
+        # AQL = Arango Query Languages
+        AQL_query = AQL_query_template
+        query_params = {"sample_ids": samples, 'num_sample_ids': len(samples)}
+        num_filters = len(filter_conditions)
+        parsed_filters = [{
+            'field': parse_field(fc.get('metadata_field'), idx),
+            'values': parse_values(fc.get('metadata_values'), idx),
+            'comp_op': parse_comparison_operator(fc.get('comparison_operator'), idx),
+            'logic_op': parse_logical_operator(fc.get('logical_operator'), idx, num_filters)
+        } for idx, fc in enumerate(filter_conditions)]
+        formatted_filters = self._validate_filters(parsed_filters)
+        for idx, formatted_filter in enumerate(formatted_filters):
+            query_constraint, filter_params = self._construct_filter(
+                formatted_filter, idx
+            )
+            logic_op = formatted_filter.get('logic_op')
+            query_params.update(filter_params)
+            if idx+1 < num_filters:
+                AQL_query += query_constraint + f" {logic_op} "
+            else:
+                # the final logical operator statement is ignored
+                AQL_query += query_constraint
+
+        AQL_query += """
+        RETURN {"id": node.id, "version": node.version}
+        """
+        results = execute_query(
+            AQL_query,
+            self.re_api_url,
+            self.re_admin_token,
+            query_params
+        )
+        return {
+            'sample_ids': results['results'],
+            'query': AQL_query
+        }
+
+    def _construct_filter(self, formatted_filter, idx):
+        '''
+        formatted_filter must contain the following parameters:
+            'field', 'comparison_operator', 'value', 'logical_operator'
+        '''
+        field = formatted_filter.get('field')
+        comp_op = formatted_filter.get('comp_op')
+        values = formatted_filter.get('values')
+        AQL_query = f"node.meta.{field}.value {comp_op} @value{idx}"
+        # if there is one value in the list of values, flatten to just the value.
+        if len(values) == 1 and comp_op not in ["IN", "NOT IN"]:
+            values = values[0]
+        filter_params = {
+            f"value{idx}": values
+        }
+        return AQL_query, filter_params
+
+    def _validate_filters(self, parsed_filters):
```

Comment:
for clarity it might be good to rename this function, or split it in two, as it seems to do more than just validate